src/github.com/matrix-org/dendrite/typingserver/api/input.go

```diff
@@ -0,0 +1,85 @@
+// Copyright 2017 Vector Creations Ltd
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package api provides the types that are used to communicate with the typing server.
+package api
+
+import (
+	"context"
+	"net/http"
+
+	commonHTTP "github.com/matrix-org/dendrite/common/http"
+	"github.com/matrix-org/gomatrixserverlib"
+	opentracing "github.com/opentracing/opentracing-go"
+)
+
+// InputTypingEvent is an event for notifying the typing server about typing updates.
+type InputTypingEvent struct {
+	// UserID of the user to update typing status.
+	UserID string `json:"user_id"`
+	// RoomID of the room the user is typing (or has stopped).
+	RoomID string `json:"room_id"`
+	// Typing is true if the user is typing, false if they have stopped.
+	Typing bool `json:"typing"`
+	// Timeout is the interval for which the user should be marked as typing.
+	Timeout int64 `json:"timeout"`
+	// OriginServerTS when the server received the update.
+	OriginServerTS gomatrixserverlib.Timestamp `json:"origin_server_ts"`
+}
+
+// InputTypingEventsRequest is a request to TypingServerInputAPI
+type InputTypingEventsRequest struct {
+	InputTypingEvents []InputTypingEvent `json:"input_typing_events"`
```

Comment:
Why do you want to input an array?
If one - but not all - requests are failing e.g. because of missing permissions for one request: How do you find out?
I think the easier way is to have one request per request. Where should the consolidating and separation happen?

Reply:
so, 1st thing is that these EDUs are already permitted (have gone through auth).
And from an perspective of drawing out an API, it seemed short-sighted to support single event per API call.

P.S. I remembered now that we'll need to put multiple events at a time for federation :P

Reply:
Given that we only currently send one typing event at a time I'd not use a slice for now. Its an internal API so we can change it down the road if we need it to handle multiple typing events :)